modules/service/src/main/resources/db/migration/V0864__emails.sql

```diff
@@ -0,0 +1,107 @@
+-- Status of an email
+CREATE TYPE e_email_status AS ENUM (
+  'queued',
+  'rejected',
+  'accepted',
+  'delivered',
+  'permanent_failure',
+  'temporary_failure'
+);
+COMMENT ON TYPE e_email_status IS 'Status of emails.';
+
+CREATE TABLE t_email (
+  c_email_id          TEXT           NOT NULL PRIMARY KEY CHECK(length(c_email_id) > 0),
+  c_program_id        d_program_id   NOT NULL REFERENCES t_program (c_program_id),
+  c_sender_email      d_email        NOT NULL,
+  c_recipient_email   d_email        NOT NULL,
+  c_subject           TEXT           NOT NULL             CHECK(length(c_subject) > 0),
+  c_text_message      TEXT           NOT NULL             CHECK(length(c_text_message) > 0),
+  c_html_message      TEXT           NULL                 CHECK(c_html_message IS NULL OR length(c_html_message) > 0),
+  c_original_time     timestamp      NOT NULL DEFAULT now(),
+  c_status            e_email_status NOT NULL,
+  c_status_time       timestamp      NOT NULL DEFAULT now()
+);
+
+-- We'll usually set this from the email webhook, but if we do it manually we'll set it to now.
+CREATE OR REPLACE FUNCTION ch_update_email_status_time()
+  RETURNS trigger AS $$
+DECLARE
+BEGIN
+  IF NEW.c_status IS DISTINCT FROM OLD.c_status AND NEW.c_status_time IS NULL THEN
+   NEW.c_status_time = now();
+  END IF;
+  return NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER ch_update_email_status_time_trigger
+  BEFORE UPDATE on t_email
+  FOR EACH ROW
+  EXECUTE PROCEDURE ch_update_email_status_time();
+
+-- Add the email id to the invitation
+
+ALTER TABLE t_invitation
+  -- Needs to be nullable because we need to add a row to get the hash before we can send the email.
+  -- Unfortunately, CHECK constraints cannot be deferred.
+  ADD COLUMN c_email_id TEXT NULL REFERENCES t_email (c_email_id);
```

Comment:
So, really, we'll always have an email id for an invitation because I roll back the transaction with the initial add if sending the email fails. But, I couldn't see a way to make it NOT NULL.